# Lab 8: Define and Solve an ML Problem of Your Choosing

In [48]:
import pandas as pd
import numpy as np
import os 
import matplotlib.pyplot as plt
import seaborn as sns

In this lab assignment, you will follow the machine learning life cycle and implement a model to solve a machine learning problem of your choosing. You will select a data set and choose a predictive problem that the data set supports.  You will then inspect the data with your problem in mind and begin to formulate a  project plan. You will then implement the machine learning project plan. 

You will complete the following tasks:

1. Build Your DataFrame
2. Define Your ML Problem
3. Perform exploratory data analysis to understand your data.
4. Define Your Project Plan
5. Implement Your Project Plan:
    * Prepare your data for your model.
    * Fit your model to the training data and evaluate your model.
    * Improve your model's performance.

## Part 1: Build Your DataFrame

You will have the option to choose one of four data sets that you have worked with in this program:

* The "census" data set that contains Census information from 1994: `censusData.csv`
* Airbnb NYC "listings" data set: `airbnbListingsData.csv`
* World Happiness Report (WHR) data set: `WHR2018Chapter2OnlineData.csv`
* Book Review data set: `bookReviewsData.csv`

Note that these are variations of the data sets that you have worked with in this program. For example, some do not include some of the preprocessing necessary for specific models. 

#### Load a Data Set and Save it as a Pandas DataFrame

The code cell below contains filenames (path + filename) for each of the four data sets available to you.

<b>Task:</b> In the code cell below, use the same method you have been using to load the data using `pd.read_csv()` and save it to DataFrame `df`. 

You can load each file as a new DataFrame to inspect the data before choosing your data set.

In [49]:
# File names of the four data sets
adultDataSet_filename = os.path.join(os.getcwd(), "data", "censusData.csv")
airbnbDataSet_filename = os.path.join(os.getcwd(), "data", "airbnbListingsData.csv")
WHRDataSet_filename = os.path.join(os.getcwd(), "data", "WHR2018Chapter2OnlineData.csv")
bookReviewDataSet_filename = os.path.join(os.getcwd(), "data", "bookReviewsData.csv")


df = pd.read_csv(WHRDataSet_filename, header=0)

df.head()

country  year  Life Ladder  Log GDP per capita  Social support  \
0  Afghanistan  2008     3.723590            7.168690        0.450662   
1  Afghanistan  2009     4.401778            7.333790        0.552308   
2  Afghanistan  2010     4.758381            7.386629        0.539075   
3  Afghanistan  2011     3.831719            7.415019        0.521104   
4  Afghanistan  2012     3.782938            7.517126        0.520637   

   Healthy life expectancy at birth  Freedom to make life choices  Generosity  \
0                         49.209663                      0.718114    0.181819   
1                         49.624432                      0.678896    0.203614   
2                         50.008961                      0.600127    0.137630   
3                         50.367298                      0.495901    0.175329   
4                         50.709263                      0.530935    0.247159   

   Perceptions of corruption  Positive affect  Negative affect  \
0                   0.881686         0.517637         0.258195   
1                   0.850035         0.583926         0.237092   
2                   0.706766         0.618265         0.275324   
3                   0.731109         0.611387         0.267175   
4                   0.775620         0.710385         0.267919   

   Confidence in national government  Democratic Quality  Delivery Quality  \
0                           0.612072           -1.929690         -1.655084   
1                           0.611545           -2.044093         -1.635025   
2                           0.299357           -1.991810         -1.617176   
3                           0.307386           -1.919018         -1.616221   
4                           0.435440           -1.842996         -1.404078   

   Standard deviation of ladder by country-year  \
0                                      1.774662   
1                                      1.722688   
2                                      1.878622   
3                                      1.785360   
4                                      1.798283   

   Standard deviation/Mean of ladder by country-year  \
0                                           0.476600   
1                                           0.391362   
2                                           0.394803   
3                                           0.465942   
4                                           0.475367   

   GINI index (World Bank estimate)  \
0                               NaN   
1                               NaN   
2                               NaN   
3                               NaN   
4                               NaN   

   GINI index (World Bank estimate), average 2000-15  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

   gini of household income reported in Gallup, by wp5-year  
0                                                NaN         
1                                           0.441906         
2                                           0.327318         
3                                           0.336764         
4                                           0.344540

## Part 2: Define Your ML Problem

Next you will formulate your ML Problem. In the markdown cell below, answer the following questions:

1. List the data set you have chosen.
2. What will you be predicting? What is the label?
3. Is this a supervised or unsupervised learning problem? Is this a clustering, classification or regression problem? Is it a binary classificaiton or multi-class classifiction problem?
4. What are your features? (note: this list may change after your explore your data)
5. Explain why this is an important problem. In other words, how would a company create value with a model that predicts this label?

1. I have chosen the World Happiness Report Data set.
2. I will be predicting the life expectancy at birth
3. This is a supervised learning and regression problem. 
4. As of now, I believe my features will be life ladder, Log GDP per capita, social support, freedom to make life choices, generosity, perceptions of corruption, positive affect, negative affect, confidence in national government, democratic quality, and delivery quality. 
5. This is an important problem because life expectancy can be an indicator of the health of the nation. This could be used to make global health decisions and understand factors relating to global health. 

## Part 3: Understand Your Data

The next step is to perform exploratory data analysis. Inspect and analyze your data set with your machine learning problem in mind. Consider the following as you inspect your data:

1. What data preparation techniques would you like to use? These data preparation techniques may include:

    * addressing missingness, such as replacing missing values with means
    * finding and replacing outliers
    * renaming features and labels
    * finding and replacing outliers
    * performing feature engineering techniques such as one-hot encoding on categorical features
    * selecting appropriate features and removing irrelevant features
    * performing specific data cleaning and preprocessing techniques for an NLP problem
    * addressing class imbalance in your data sample to promote fair AI
    

2. What machine learning model (or models) you would like to use that is suitable for your predictive problem and data?
    * Are there other data preparation techniques that you will need to apply to build a balanced modeling data set for your problem and model? For example, will you need to scale your data?
 
 
3. How will you evaluate and improve the model's performance?
    * Are there specific evaluation metrics and methods that are appropriate for your model?
    

Think of the different techniques you have used to inspect and analyze your data in this course. These include using Pandas to apply data filters, using the Pandas `describe()` method to get insight into key statistics for each column, using the Pandas `dtypes` property to inspect the data type of each column, and using Matplotlib and Seaborn to detect outliers and visualize relationships between features and labels. If you are working on a classification problem, use techniques you have learned to determine if there is class imbalance.

<b>Task</b>: Use the techniques you have learned in this course to inspect and analyze your data. You can import additional packages that you have used in this course that you will need to perform this task.

<b>Note</b>: You can add code cells if needed by going to the <b>Insert</b> menu and clicking on <b>Insert Cell Below</b> in the drop-drown menu.

In [50]:
df.shape

(1562, 19)

In [51]:
df_summ = df.describe()
df_summ

year  Life Ladder  Log GDP per capita  Social support  \
count  1562.000000  1562.000000         1535.000000     1549.000000   
mean   2011.820743     5.433676            9.220822        0.810669   
std       3.419787     1.121017            1.184035        0.119370   
min    2005.000000     2.661718            6.377396        0.290184   
25%    2009.000000     4.606351            8.310665        0.748304   
50%    2012.000000     5.332600            9.398610        0.833047   
75%    2015.000000     6.271025           10.190634        0.904329   
max    2017.000000     8.018934           11.770276        0.987343   

       Healthy life expectancy at birth  Freedom to make life choices  \
count                       1553.000000                   1533.000000   
mean                          62.249887                      0.728975   
std                            7.960671                      0.145408   
min                           37.766476                      0.257534   
25%                           57.299580                      0.633754   
50%                           63.803192                      0.748014   
75%                           68.098228                      0.843628   
max                           76.536362                      0.985178   

        Generosity  Perceptions of corruption  Positive affect  \
count  1482.000000                1472.000000      1544.000000   
mean      0.000079                   0.753622         0.708969   
std       0.164202                   0.185538         0.107644   
min      -0.322952                   0.035198         0.362498   
25%      -0.114313                   0.697359         0.621471   
50%      -0.022638                   0.808115         0.717398   
75%       0.094649                   0.880089         0.800858   
max       0.677773                   0.983276         0.943621   

       Negative affect  Confidence in national government  Democratic Quality  \
count      1550.000000                        1401.000000         1391.000000   
mean          0.263171                           0.480207           -0.126617   
std           0.084006                           0.190724            0.873259   
min           0.083426                           0.068769           -2.448228   
25%           0.204116                           0.334732           -0.772010   
50%           0.251798                           0.463137           -0.225939   
75%           0.311515                           0.610723            0.665944   
max           0.704590                           0.993604            1.540097   

       Delivery Quality  Standard deviation of ladder by country-year  \
count       1391.000000                                   1562.000000   
mean           0.004947                                      2.003501   
std            0.981052                                      0.379684   
min           -2.144974                                      0.863034   
25%           -0.717463                                      1.737934   
50%           -0.210142                                      1.960345   
75%            0.717996                                      2.215920   
max            2.184725                                      3.527820   

       Standard deviation/Mean of ladder by country-year  \
count                                        1562.000000   
mean                                            0.387271   
std                                             0.119007   
min                                             0.133908   
25%                                             0.309722   
50%                                             0.369751   
75%                                             0.451833   
max                                             1.022769   

       GINI index (World Bank estimate)  \
count                        583.000000   
mean                           0.372846   
std                            0.086609   
min                 

In [52]:
#Getting relevant statistics from data. 
#One insight we can make is which features have a heavy skew (varying median and mean)
df_summ_all = df.describe(include = 'all')
df_summ_all

country         year  Life Ladder  Log GDP per capita  \
count       1562  1562.000000  1562.000000         1535.000000   
unique       164          NaN          NaN                 NaN   
top     Zimbabwe          NaN          NaN                 NaN   
freq          12          NaN          NaN                 NaN   
mean         NaN  2011.820743     5.433676            9.220822   
std          NaN     3.419787     1.121017            1.184035   
min          NaN  2005.000000     2.661718            6.377396   
25%          NaN  2009.000000     4.606351            8.310665   
50%          NaN  2012.000000     5.332600            9.398610   
75%          NaN  2015.000000     6.271025           10.190634   
max          NaN  2017.000000     8.018934           11.770276   

        Social support  Healthy life expectancy at birth  \
count      1549.000000                       1553.000000   
unique             NaN                               NaN   
top                NaN                               NaN   
freq               NaN                               NaN   
mean          0.810669                         62.249887   
std           0.119370                          7.960671   
min           0.290184                         37.766476   
25%           0.748304                         57.299580   
50%           0.833047                         63.803192   
75%           0.904329                         68.098228   
max           0.987343                         76.536362   

        Freedom to make life choices   Generosity  Perceptions of corruption  \
count                    1533.000000  1482.000000                1472.000000   
unique                           NaN          NaN                        NaN   
top                              NaN          NaN                        NaN   
freq                             NaN          NaN                        NaN   
mean                        0.728975     0.000079                   0.753622   
std                         0.145408     0.164202                   0.185538   
min                         0.257534    -0.322952                   0.035198   
25%                         0.633754    -0.114313                   0.697359   
50%                         0.748014    -0.022638                   0.808115   
75%                         0.843628     0.094649                   0.880089   
max                         0.985178     0.677773                   0.983276   

        Positive affect  Negative affect  Confidence in national government  \
count       1544.000000      1550.000000                        1401.000000   
unique              NaN              NaN                                NaN   
top                 NaN              NaN                                NaN   
freq                NaN              NaN                                NaN   
mean           0.708969         0.263171                           0.480207   
std            0.107644         0.084006                           0.190724   
min            0.362498         0.083426                           0.068769   
25%            0.621471         0.204116                           0.334732   
50%            0.717398         0.251798                           0.463137   
75%            0.800858         0.311515                           0.610723   
max            0.943621         0.704590                           0.993604   

        Democratic Quality  Delivery Quality  \
count          1391.000000       1391.000000   
unique                 NaN               NaN   
top                    NaN               NaN   
freq                   NaN               NaN   
mean             -0.126617          0.004947   
std               0.873259          0.981052   
min              -2.448228         -2.144974   
25%              -0.772010         -0.717463   
50%              -0.225939         -0.210142   
75%               0.665944          0.717996   
max               1.540097          2.184725   

        

In [53]:
#Examining the correlation coefficient between label and features and sorting them. We can use this data to see
#which features have the highest correlation with healthy life expectancy at birth 
corrs = df.corr()['Healthy life expectancy at birth']
corrs_sorted = corrs.sort_values(ascending=False)
corrs_sorted

Healthy life expectancy at birth                            1.000000
Log GDP per capita                                          0.841612
Life Ladder                                                 0.729852
Delivery Quality                                            0.721081
Democratic Quality                                          0.597106
Social support                                              0.586759
Freedom to make life choices                                0.340026
Positive affect                                             0.297759
year                                                        0.100904
Generosity                                                  0.047079
Standard deviation of ladder by country-year               -0.065870
Negative affect                                            -0.105255
Confidence in national government                          -0.188827
GINI index (World Bank estimate)                           -0.306798
Perceptions of corruption         

In [54]:
#Examining null value counts for features. Features with high amounts of null values may not be desired.
nan_count = np.sum(df.isnull(), axis = 0)
print(nan_count)

country                                                       0
year                                                          0
Life Ladder                                                   0
Log GDP per capita                                           27
Social support                                               13
Healthy life expectancy at birth                              9
Freedom to make life choices                                 29
Generosity                                                   80
Perceptions of corruption                                    90
Positive affect                                              18
Negative affect                                              12
Confidence in national government                           161
Democratic Quality                                          171
Delivery Quality                                            171
Standard deviation of ladder by country-year                  0
Standard deviation/Mean of ladder by cou

## Part 4: Define Your Project Plan

Now that you understand your data, in the markdown cell below, define your plan to implement the remaining phases of the machine learning life cycle (data preparation, modeling, evaluation) to solve your ML problem. Answer the following questions:

* Do you have a new feature list? If so, what are the features that you chose to keep and remove after inspecting the data? 
* Explain different data preparation techniques that you will use to prepare your data for modeling.
* What is your model (or models)?
* Describe your plan to train your model, analyze its performance and then improve the model. That is, describe your model building, validation and selection plan to produce a model that generalizes well to new data. 

Yes: I removed freedom to make life choices, generosity, perceptions of corruption, positive affect, negative affect, and confidence in national government because they had low correlation with the label. I will replace null values with means in order to prepare my data, as well as seperate my data into training and test sets. I will be using a linear regression model. I will fit the linear regression model to the training data and then make predictions on test data. I will examine the 

## Part 5: Implement Your Project Plan

<b>Task:</b> In the code cell below, import additional packages that you have used in this course that you will need to implement your project plan.

In [55]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

<b>Task:</b> Use the rest of this notebook to carry out your project plan. 

You will:

1. Prepare your data for your model.
2. Fit your model to the training data and evaluate your model.
3. Improve your model's performance by performing model selection and/or feature selection techniques to find best model for your problem.

Add code cells below and populate the notebook with commentary, code, analyses, results, and figures as you see fit. 

In [82]:
#Initially, null values were simply removed. However, employing the method of using the means of the features to 
#replace null values yielded a higher r^2 score - and was thus kept. 
mean_GDP=df['Log GDP per capita'].mean()
df['Log GDP per capita'].fillna(value=mean_GDP, inplace=True)
mean_SS=df['Social support'].mean()
df['Social support'].fillna(value=mean_SS, inplace=True)
mean_demQ=df['Democratic Quality'].mean()
df['Democratic Quality'].fillna(value=mean_demQ, inplace=True)
mean_delQ=df['Delivery Quality'].mean()
df['Delivery Quality'].fillna(value=mean_delQ, inplace=True)
mean_lifeExp=df['Healthy life expectancy at birth'].mean()
df['Healthy life expectancy at birth'].fillna(value=mean_lifeExp, inplace=True)

In [84]:
nan_count = np.sum(df[features].isnull(), axis = 0)
nan_count_label = np.sum(df['Healthy life expectancy at birth'].isnull(), axis = 0)

print(nan_count)
print(nan_count_label)

Life Ladder           0
Log GDP per capita    0
Social support        0
Delivery Quality      0
Democratic Quality    0
dtype: int64
0


In [85]:
features = ['Life Ladder', 'Log GDP per capita', 'Social support', 'Delivery Quality', 'Democratic Quality']
X = df[features]
X.head()
y = df['Healthy life expectancy at birth']
print(X)

      Life Ladder  Log GDP per capita  Social support  Delivery Quality  \
0        3.723590            7.168690        0.450662         -1.655084   
1        4.401778            7.333790        0.552308         -1.635025   
2        4.758381            7.386629        0.539075         -1.617176   
3        3.831719            7.415019        0.521104         -1.616221   
4        3.782938            7.517126        0.520637         -1.404078   
...           ...                 ...             ...               ...   
1557     4.690188            7.565154        0.799274         -1.526321   
1558     4.184451            7.562753        0.765839         -1.484067   
1559     3.703191            7.556052        0.735800         -1.357514   
1560     3.735400            7.538829        0.768425         -1.371214   
1561     3.638300            7.538187        0.754147          0.004947   

      Democratic Quality  
0              -1.929690  
1              -2.044093  
2              -1.

In [86]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.10, random_state=1234)

In [87]:
# Create the  LinearRegression model object 
model = LinearRegression()

# Fit the model to the training data 
model.fit(X_train, y_train)

#  Make predictions on the test data 
prediction = model.predict(X_test)

In [88]:
print('Model Summary:\n')

# Print intercept (alpha)
print('Intercept:')
print('alpha = ' , model.intercept_)

# Print weights
print('\nWeights:')
i = 0
for w in model.coef_:
    print('w_',i+1,'= ', w, ' [ weight of ', features[i],']')
    i += 1

Model Summary:

Intercept:
alpha =  18.200172241085156

Weights:
w_ 1 =  1.1566145692864906  [ weight of  Life Ladder ]
w_ 2 =  4.008174687480497  [ weight of  Log GDP per capita ]
w_ 3 =  0.9695817705723393  [ weight of  Social support ]
w_ 4 =  1.2324010450081953  [ weight of  Delivery Quality ]
w_ 5 =  -0.04407505757833907  [ weight of  Democratic Quality ]


In [89]:
# Print mean squared error
print('\nModel Performance\n\nRMSE =   %.2f'
      % np.sqrt(mean_squared_error(y_test, prediction)))
# The coefficient of determination
print(' R^2 =   %.2f'
      % r2_score(y_test, prediction))


Model Performance

RMSE =   3.85
 R^2 =   0.76


The RMSE value of 3.85 and R^2 value of 0.76 indicates that the regression model fits the data fairly well. 